pkg/config/reader.go

```diff
@@ -128,6 +135,105 @@ func (r *FileReader) validateConfig() error {
 	return nil
 }
 
+func (r *FileReader) mergePresets(v *viper.Viper) error {
+	if len(r.cfg.Presets) == 0 {
+		return nil
+	}
+	presets, err := r.loadPresets(v)
+	if err != nil {
+		return err
+	}
+
+	// Merge via viper, with special handling for .linters.{en,dis}abled slice
+	mergedV := viper.New()
+	lintersEnabled := map[string]struct{}{}
+	lintersDisabled := map[string]struct{}{}
+	for _, cfg := range append(presets, v) {
+		if err := mergedV.MergeConfigMap(cfg.AllSettings()); err != nil {
+			return fmt.Errorf("can't merge config %q: %w", cfg.ConfigFileUsed(), err)
+		}
+
+		for _, l := range cfg.GetStringSlice("linters.enable") {
+			lintersEnabled[l] = struct{}{}
+			delete(lintersDisabled, l)
+		}
+		for _, l := range cfg.GetStringSlice("linters.disable") {
+			lintersDisabled[l] = struct{}{}
+			delete(lintersEnabled, l)
+		}
+	}
+
+	if err := mergedV.Unmarshal(r.cfg); err != nil {
+		return fmt.Errorf("can't unmarshal merged config: %w", err)
+	}
+
+	if len(lintersEnabled) > 0 {
+		r.cfg.Linters.Enable = make([]string, 0, len(lintersEnabled))
+		for l := range lintersEnabled {
+			r.cfg.Linters.Enable = append(r.cfg.Linters.Enable, l)
+		}
+		sort.Strings(r.cfg.Linters.Enable)
+	} else {
+		r.cfg.Linters.Enable = nil
+	}
+	if len(lintersDisabled) > 0 {
+		r.cfg.Linters.Disable = make([]string, 0, len(lintersDisabled))
+		for l := range lintersDisabled {
+			r.cfg.Linters.Disable = append(r.cfg.Linters.Disable, l)
+		}
+		sort.Strings(r.cfg.Linters.Disable)
+	} else {
+		r.cfg.Linters.Disable = nil
+	}
+
+	return nil
+}
+
+const configName = ".golangci"
+
+func (r *FileReader) loadPresets(v *viper.Viper) ([]*viper.Viper, error) {
+	cfgBase := filepath.Dir(v.ConfigFileUsed())
+	presets := make([]*viper.Viper, 0, len(r.cfg.Presets))
+	for _, preset := range r.cfg.Presets {
+		presetV := viper.New()
+		if strings.HasPrefix(preset, ".") {
```

Comment:
It means the local path always should be started with `./` ?

Reply:
Or `../`. The pattern for distinguishing modules from local paths is inspired by the [right hand side of replace directives](https://golang.org/ref/mod#go-mod-file-replace) in `go.mod`.

> If the path on the right side of the arrow is an absolute or relative path (beginning with ./ or ../), it is interpreted as the local file path to the replacement module root directory, which must contain a go.mod file.
> [...]
> If the path on the right side is not a local path, it must be a valid module path.

`/` might also be a valid prefix to trigger file loading that I ~missed~ did not think would be used.